libraries/Microsoft.Bot.Builder/BotFrameworkStreamingExtensionsAdapter.cs

```diff
@@ -0,0 +1,695 @@
+            }
+
+            if (callback == null)
+            {
+                throw new ArgumentNullException(nameof(callback));
+            }
+
+            if (cancellationToken == null)
+            {
+                throw new ArgumentNullException(nameof(cancellationToken));
+            }
+
+            var activity = JsonConvert.DeserializeObject<Activity>(body, SerializationSettings.DefaultDeserializationSettings);
+
+            if (streams.Count > 1)
+            {
+                var streamAttachments = new List<Attachment>();
+                for (var i = 1; i < streams.Count; i++)
+                {
+                    streamAttachments.Add(new Attachment() { ContentType = streams[i].Type, Content = streams[i].GetStream() });
+                }
+
+                if (activity.Attachments != null)
+                {
+                    activity.Attachments = activity.Attachments.Concat(streamAttachments).ToArray();
+                }
+                else
+                {
+                    activity.Attachments = streamAttachments.ToArray();
+                }
+            }
+
+            return await ProcessActivityAsync(activity, callback, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Primary adapter method for processing activities sent from channel.
+        /// Creates a turn context and runs the middleware pipeline for an incoming activity.
+        /// Throws <see cref="ArgumentNullException"/> on null arguments.
+        /// </summary>
+        /// <param name="activity">The <see cref="Activity"/> to process.</param>
+        /// <param name="callback">The code to run at the end of the adapter's middleware pipeline.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute. If the activity type
+        /// was 'Invoke' and the corresponding key (channelId + activityId) was found
+        /// then an InvokeResponse is returned, otherwise null is returned.</returns> /// <returns>a <see cref="Task"/> that will resolve a <see cref="InvokeResponse"/> in response to the activity.</returns>
+        /// <remarks>Call this method to reactively send a message to a conversation.
+        /// If the task completes successfully, then if the activity's <see cref="Activity.Type"/>
+        /// is <see cref="ActivityTypes.Invoke"/> and the corresponding key
+        /// (<see cref="Activity.ChannelId"/> + <see cref="Activity.Id"/>) is found
+        /// then an <see cref="InvokeResponse"/> is returned, otherwise null is returned.
+        /// <para>This method registers the following services for the turn.<list type="bullet">
+        public async Task<InvokeResponse> ProcessActivityAsync(Activity activity, BotCallbackHandler callback, CancellationToken cancellationToken)
+        {
```

Comment:
This whole code block (indeed, most of this class) looks like a cut/paste from the Http adapter. Can we refactor the code in a bit to reflect that fact?

Reply:
We've discussed a few promising approaches to refactoring this entire class, I'd like to hold off until we can address it all at once.